diff --git a/shell/platform/android/android_shell_holder.cc b/shell/platform/android/android_shell_holder.cc
index 1e5b9e95c0..7b24bb9c40 100644
--- a/shell/platform/android/android_shell_holder.cc
+++ b/shell/platform/android/android_shell_holder.cc
@@ -16,6 +16,7 @@
 #include <string>
 #include <utility>
 
+#include "flutter/assets/directory_asset_bundle.h"
 #include "flutter/fml/logging.h"
 #include "flutter/fml/make_copyable.h"
 #include "flutter/fml/message_loop.h"
@@ -331,6 +332,16 @@ std::optional<RunConfiguration> AndroidShellHolder::BuildRunConfiguration(
   }
 
   RunConfiguration config(std::move(isolate_configuration));
+  
+  auto asset_dir = GetSettings().assets_path;
+  if (!asset_dir.empty()) {
+    auto dir_asset_provider = std::make_unique<DirectoryAssetBundle>(
+      fml::OpenDirectory(asset_dir.c_str(), false,
+                         fml::FilePermission::kRead),
+      false);
+    config.AddAssetResolver(std::move(dir_asset_provider));
+  }
+
   config.AddAssetResolver(apk_asset_provider_->Clone());
 
   {
