diff --git a/shell/platform/darwin/ios/platform_message_handler_ios.mm b/shell/platform/darwin/ios/platform_message_handler_ios.mm
index cf510bc..a5a2b15 100644
--- a/shell/platform/darwin/ios/platform_message_handler_ios.mm
+++ b/shell/platform/darwin/ios/platform_message_handler_ios.mm
@@ -16,7 +16,7 @@ static uint64_t platform_message_counter = 1;
 @end
 
 @interface FLTSerialTaskQueue : NSObject <FlutterTaskQueue>
-@property(nonatomic, strong) dispatch_queue_t queue;
+@property(nonatomic, assign) dispatch_queue_t queue;
 @end
 
 @implementation FLTSerialTaskQueue
